text/0000-default-field-values.md

```diff
@@ -0,0 +1,1802 @@
+it had been attached to a `struct` with the same fields and field visibility.
+
+### Interaction with `#[default]`
+
+[`#[default]`]: https://github.com/rust-lang/rfcs/pull/3107
+
+It is possible today to specify a [`#[default]` variant] in an enum so that it
+can be `#[derive(Default)]`. A variant marked with `#[default]` will use
+defaulted fields when present.
+
+```rust
+#[derive(Default)]
+enum Ingredient {
+    Tomato {
+        color: Color = Color::Red,
+        taste: TasteQuality,
+    },
+    Onion {
+        color: Color = Color::Yellow,
+    },
+    #[default]
+    Lettuce {
+        color: Color = Color::Green,
+    },
+}
+```
+
+Now the compiler does know that `Ingredient::Lettuce` should be considered
+the default and will accordingly generate an appropriate implementation of
+`Default for Ingredient` (19):
+
+```rust
+impl Default for Ingredient {
+    fn default() -> Self {
+        Ingredient::Lettuce {
+            color: Color::Green,
+        }
+    }
+}
+```
+
+## Defaults on tuple `struct`s and tuple `enum` variants
+
+Default values are only allowed on named fields. There is no syntax provided for
+tuple types like `struct S(i32)` or `enum E { V(i32), }`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Field default values
+
+### Grammar
+
+Let the grammar of record fields in `struct`s and `enum` variants be defined
+like so (in the `.lyg` notation):
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type;
+```
+
+Then, `RecordField` is changed into:
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type { "=" def:Expr }?;
+```
+
+Further, given the following partial definition for the expression grammar:
+
+```rust
+Expr = attrs:OuterAttr* kind:ExprKind;
+ExprKind =
+  | ...
+  | Struct:{ path:Path "{" attrs:InnerAttr* fields:StructExprFieldsAndBase "}" }
+  ;
+
+StructExprFieldsAndBase =
+  | Fields:{ fields:StructExprField* % "," ","? }
+  | Base:{ ".." base:Expr }
+  | FieldsAndBase:{ fields:StructExprField+ % "," "," ".." base:Expr }
+  ;
+StructExprField = attrs:OuterAttr* kind:StructExprFieldKind;
+StructExprFieldKind =
+  | Shorthand:IDENT
+  | Explicit:{ field:FieldName ":" expr:Expr }
+  ;
+```
+
+the rule `StructExprFieldsAndBase` is extended with:
+
+```rust
+StructExprFieldsAndBase =| FieldsAndDefault:{ fields:StructExprField+ % "," "," ".." };
+StructExprFieldsAndBase =| Default:{ ".." }
+```
+
+### Static semantics
+
+#### Defining defaults
+
+Given a `RecordField` where the default is specified, i.e.:
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type "=" def:Expr;
+```
+
+all the following rules apply when type-checking:
+
+1. The expression `def` must be a constant expression.
+
+2. The expression `def` must coerce to the type `ty`.
+
+When lints check attributes such as `#[allow(lint_name)]` are placed on a
+`RecordField`, it also applies to `def` if it exists.
```

Comment:
Are the default expressions able to make use of generics, including those defined in function scope? E.g., if I have:

```rust
fn foo<const A: usize>() {
    struct Bar<const B: usize> {
        field1: usize = A,
        field2: usize = B,
    }
}
```

Are both fields valid, or neither? In some sense, this is asking whether the desugaring is to a free-standing `const` item whose value is referenced, or an inline `const` block.

It's probably also worth asking/defining when the default expressions are evaluated -- e.g., if the expression panics, does that prevent compilation of that struct definition? Presuming they can reference generics, can I define a struct that doesn't use that constant (e.g., in the example above, if we explicitly initialize the struct in all use sites with both fields, does that mean that the struct can have default values that always/conditionally panic?)

Reply:
Current output for this on my draft impl:

```
error[E0401]: can't use generic parameters from outer item
  --> f80.rs:51:25
   |
49 | fn qux<const A: usize>() {
   |              - const parameter from outer item
50 |     struct Bar<const B: usize> {
   |                - help: try introducing a local generic parameter here: `A,`
51 |         field1: usize = A,
   |                         ^ use of generic parameter from outer item

error: internal compiler error: compiler/rustc_middle/src/ty/consts.rs:326:13: try_from_lit: received const param which shouldn't be possible
  --> f80.rs:52:25
   |
52 |         field2: usize = B,
   |                         ^

thread 'rustc' panicked at compiler/rustc_middle/src/ty/consts.rs:326:13:
```

I think that `B` *should* be allowed, while `A` isn't, which is what we do already in other cases for items.

Reply:
That seems reasonable to me (i.e., treat as const items).

What about the evaluation side of things?

Reply:
Right, the tricky bit (IMO) is for the case `x = Config { height: 1, .. }` -- what is the MIR for that?

```
_0 = Config{ height: { const Config::__DEFAULT_FOR_HEIGHT; 1 }, width: const Config::__DEFAULT_FOR_WIDTH };
```

I think it probably ought to be something like this, where we *still evaluate* the defaults regardless of actually using them. But in any case, that needs specifying.

Reply:
Funnily enough, the current output for your examples (after cleanup):

1. success
2.
```
error[E0277]: the size for values of type `T` cannot be known at compilation time
   --> f81.rs:6:21
    |
5   |     struct Foo<T> {
    |                - this type parameter needs to be `Sized`
6   |         v: Vec<T> = Vec::new(),
    |                     ^^^^^^^^^^ doesn't have a size known at compile-time
    |
note: required by an implicit `Sized` bound in `Vec`
   --> /home/gh-estebank/rust/library/alloc/src/vec/mod.rs:397:16
    |
397 | pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
    |                ^ required by the implicit `Sized` requirement on this type parameter in `Vec`

error[E0277]: the size for values of type `T` cannot be known at compilation time
   --> f81.rs:6:21
    |
5   |     struct Foo<T> {
    |                - this type parameter needs to be `Sized`
6   |         v: Vec<T> = Vec::new(),
    |                     ^^^ doesn't have a size known at compile-time
    |
note: required by an implicit `Sized` bound in `Vec`
   --> /home/gh-estebank/rust/library/alloc/src/vec/mod.rs:397:16
    |
397 | pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
    |                ^ required by the implicit `Sized` requirement on this type parameter in `Vec`
```
3. Needs `#![feature(generic_const_exprs)]`, works as expected, and detects underflow when constructing the value:
```
error[E0080]: evaluation of `S::<0>::a::{constant#0}` failed
 --> f81.rs:5:14
  |
5 |     a: u32 = A - 1,
  |              ^^^^^ attempt to compute `0_u32 - 1_u32`, which would overflow
```
```

Reply:
Now the last case (`Vec<T>`) [now works](https://github.com/rust-lang/rust/pull/129514/commits/14777ce5eb88e76047ace5144b4ae1111fcbce1f).